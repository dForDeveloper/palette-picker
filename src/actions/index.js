export const setColors = (colors) => ({
  type: 'SET_COLORS',
  colors
});

export const setLockedColors = (lockedColors) => ({
  type: 'SET_LOCKED_COLORS',
  lockedColors
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  bool
});

export const setError = (message) => ({
  type: 'SET_ERROR',
  message
});

export const setPalettes = (id, palettes) => ({
  type: 'SET_PALETTES',
  id,
  palettes
});

export const setModal = (isDisplayed, modalType = null, text = null) => ({
  type: 'SET_MODAL',
  isDisplayed,
  modalType,
  text
});