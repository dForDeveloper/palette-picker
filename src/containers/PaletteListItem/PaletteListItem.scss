.Menu--palette-name {
  @extend .Menu--text;
}

.Menu--palette-span {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  border: 1px solid #ffffff;
  margin-top: 2px;
  margin-bottom: 8px;
  cursor: pointer;
}

.Menu--palette-square {
  width: 100%;
  height: 16px;
  display: inline-block;
}