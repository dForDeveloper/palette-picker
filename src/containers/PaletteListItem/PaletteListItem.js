import React, { Component } from 'react';
import { setColors, setLockedColors, setModal } from '../../actions';
import { connect } from 'react-redux';
import editlight from '../../icons/editlight.svg';
import PropTypes from 'prop-types';

export class PaletteListItem extends Component {
  setModal = () => {
    this.props.setModal(true, 'palette', this.props.name);
  }

  renderPaletteRectangle = (colors) => {
    return colors.map(color => {
      return (
        <div
          style={{ backgroundColor: color }}
          className="Menu--palette-square"
          key={color}
        ></div>
      );
    });
  }

  render() {
    const { name, color1, color2, color3, color4, color5 } = this.props;
    const colors = [color1, color2, color3, color4, color5];
    return (
      <li>
        <p className="Menu--palette-name">
          {name}
        </p>
        <img
          src={editlight}
          alt="edit icon"
          className="Menu--icon-edit-light"
          onClick={this.setModal}
        />
        <span
          className="Menu--palette-span"
          onClick={() => {
            this.props.setColors(colors);
            this.props.setLockedColors([]);
          }}
        >
          {this.renderPaletteRectangle(colors)}
        </span>
      </li>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  setColors: (colors) => dispatch(setColors(colors)),
  setLockedColors: (lockedColors) => dispatch(setLockedColors(lockedColors)),
  setModal: (isDisplayed, type, text) => {
    dispatch(setModal(isDisplayed, type, text))
  }
});

export default connect(null, mapDispatchToProps)(PaletteListItem);

PaletteListItem.propTypes = {
  color1: PropTypes.string,
  color2: PropTypes.string,
  color3: PropTypes.string,
  color4: PropTypes.string,
  color5: PropTypes.string,
  created_at: PropTypes.string,
  id: PropTypes.number,
  name: PropTypes.string,
  project_id:  PropTypes.number,
  updated_at: PropTypes.string,
  setColors: PropTypes.func,
  setLockedColors: PropTypes.func
};