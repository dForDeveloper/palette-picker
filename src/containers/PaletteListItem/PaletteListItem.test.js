import React from 'react';
import { shallow } from 'enzyme';
import { setColors, setLockedColors, setModal } from '../../actions';
import { PaletteListItem, mapDispatchToProps } from './PaletteListItem';

const mockColors = ['#123456', '#234567', '#345678', '#456789', '#56789a'];

describe('PaletteListItem', () => {
  let wrapper;
  const mockProps = {
    color1: '#123456',
    color2: '#234567',
    color3: '#345678',
    color4: '#456789',
    color5: '#56789a',
    id: 1,
    name: 'my palette',
    project_id:  1,
    setColors: jest.fn(),
    setLockedColors: jest.fn(),
    setModal: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<PaletteListItem {...mockProps} />);
  });

  describe('render', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call setModal when the edit icon is clicked', () => {
      jest.spyOn(wrapper.instance(), 'setModal');
      wrapper.instance().forceUpdate();
      wrapper.find('.Menu--icon-edit-light').simulate('click');
      expect(wrapper.instance().setModal).toHaveBeenCalled();
    });

    it('should call setColors when the span is clicked', () => {
      wrapper.find('.Menu--palette-span').simulate('click');
      expect(mockProps.setColors).toHaveBeenCalled();
    });
    
    it('should call setLockedColors when the span is clicked', () => {
      wrapper.find('.Menu--palette-span').simulate('click');
      expect(mockProps.setLockedColors).toHaveBeenCalled();
    });
  });

  describe('setModal', () => {
    it('should call setModal', () => {
      wrapper.instance().setModal();
      expect(mockProps.setModal)
      .toHaveBeenCalledWith(true, 'palette', 'my palette', 1, 1);
    });
  });

  describe('renderPaletteRectangle', () => {
    it('should return an array of five divs', () => {
      const result = wrapper.instance().renderPaletteRectangle(mockColors);
      expect(result).toHaveLength(5);
    });
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mappedProps = mapDispatchToProps(mockDispatch);

  it('should call setColors', () => {
    const expected = setColors(mockColors);
    mappedProps.setColors(mockColors);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
  
  it('should call setLockedColors', () => {
    const expected = setLockedColors([]);
    mappedProps.setLockedColors([]);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
  
  it('should call setModal', () => {
    const expected = setModal(true, 'palette', 'my palette', 1, 1);
    mappedProps.setModal(true, 'palette', 'my palette', 1, 1);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});