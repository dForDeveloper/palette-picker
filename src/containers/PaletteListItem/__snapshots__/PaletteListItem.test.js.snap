// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteListItem render should match the snapshot 1`] = `
<li>
  <p
    className="Menu--palette-name"
  >
    my palette
  </p>
  <img
    alt="edit icon"
    className="Menu--icon-edit-light"
    onClick={[Function]}
    src="editlight.svg"
  />
  <span
    className="Menu--palette-span"
    onClick={[Function]}
  >
    <div
      className="Menu--palette-square"
      key="#123456"
      style={
        Object {
          "backgroundColor": "#123456",
        }
      }
    />
    <div
      className="Menu--palette-square"
      key="#234567"
      style={
        Object {
          "backgroundColor": "#234567",
        }
      }
    />
    <div
      className="Menu--palette-square"
      key="#345678"
      style={
        Object {
          "backgroundColor": "#345678",
        }
      }
    />
    <div
      className="Menu--palette-square"
      key="#456789"
      style={
        Object {
          "backgroundColor": "#456789",
        }
      }
    />
    <div
      className="Menu--palette-square"
      key="#56789a"
      style={
        Object {
          "backgroundColor": "#56789a",
        }
      }
    />
  </span>
</li>
`;
